Dockerfile

```diff
@@ -1 +1,41 @@
+ARG ARCH="amd64"
+ARG OS="linux"
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base:latest
+LABEL maintainer="Armedia, LLC"
 
+ARG ARCH="amd64"
+ARG OS="linux"
+ARG VER="2.28.1"
+ARG PKG="prometheus"
+ARG SRC="${PKG}-${VER}.${OS}-${ARCH}"
+
+RUN curl \
+		-L "https://github.com/prometheus/${PKG}/releases/download/v${VER}/${SRC}.tar.gz" \
+		-o package.tar.gz && \
+	tar -xzvf "package.tar.gz" && \
+	mkdir -pv "/prometheus" && \
+	mkdir -pv "/etc/prometheus" && \
+	mkdir -pv "/usr/share/prometheus" && \
+	mv -vif "${SRC}/LICENSE"                     "/LICENSE" && \
+	mv -vif "${SRC}/NOTICE"                      "/NOTICE" && \
+	mv -vif "${SRC}/prometheus"                  "/bin/prometheus" && \
+	mv -vif "${SRC}/promtool"                    "/bin/promtool" && \
+	mv -vif "${SRC}/prometheus.yml"              "/etc/prometheus/prometheus.yml" && \
+	mv -vif "${SRC}/console_libraries/"          "/usr/share/prometheus/" && \
+	mv -vif "${SRC}/consoles/"                   "/usr/share/prometheus/" && \
+	ln -s "/usr/share/prometheus/console_libraries" "/etc/prometheus" && \
+	ln -s "/usr/share/prometheus/consoles"          "/etc/prometheus" && \
+	chown -R nobody:nobody "/etc/prometheus" "/prometheus" && \
```

Comment:
I would prefer to see `root:root` here, as it is more restrictive. But we had some discussions about this and it seems that Armedia is OK to be not very tight on file/directory permissions. So just take this as a comment/suggestion.

Reply:
The value of using an exact user/group here is that then you can grant extra-container access to container files (in mounted volumes, for instance) without requiring root access.

Reply:
@drivera-armedia I don't understand. Could you please expand or provide some examples?

Reply:
This is about mounted volumes. If I mount a directory unto a container, its contents will have to have permissions such that a) the running process can read/write as appropriate to those locations, and b) a user outside in the host machine can read/write as appropriate to those locations.

If those directories are owned by root (UID 0), then the process inside the container must run as root. If the process were compromised, then that process could do a LOT of damage within the container. While the container ... _contains_ the damage, you have a 2nd problem: to access the data stored within those mounted volumes, you'd have to be root yourself which tends to be a very bad thing to spread around.

So, by using a lowered user we can maintain isolation while also gaining access as needed, but without needing to be root or resorting to granting global read/write/execute permissions.

Reply:
@drivera-armedia That seems a bit obscure to me... Generally speaking, I think anything that must be writeable by the containerized app should be isolated in separate directories. For example with ActiveMQ, I have `/app/conf` for the configuration files which all owned by root and read-only for the `activemq` user. Ditto for other files/directories that ActiveMQ needs to be able to access only on a read-only basis. For files/directories that ActiveMQ needs to be able to write to, that's under `/app/data`, which is owned by the `activemq` user and with permissions to write. When the container is started, Kubernetes will mount a persistent volume on `/app/data`. 

In your example, I still don't understand why, for example, `/etc/prometheus` should be owned by `nobody`. I Prometheus will not write into this directory (at least I think it should not). I guess you will have only static configuration files in there. Are saying that you want to mount a volume on `/etc/prometheus`? Even if that's the case, I don't understand why it should be owned by `nobody`...

Reply:
Nobody was used out of the practicality of using an existing user without having to create one. At the O/S level there's no difference between using UID X or UID Y as long as neither is "0", and neither clashes with an existing host system user (i.e. 501).

There's no problem in creating a user for prometheus with a given UID, except the aforementioned potential for collision with an existing host O/S user which would grant access to someone not meant to gain that access.  If we define a UID range to use for creating our users, then definitely this is something to consider ... but now we would have to maintain a list of users.

Say we deploy Prometheus alongside ActiveMQ in the same machine. It just so happens that when the containers' **prometheus** and **activemq** users were created, they both got the same UID ... and it just so happens that there's a user in the host system with that UID already. This means that this outside user already - albeit likely unknowingly - has access to both Pometheus's and ActiveMQ's data volumes...

Granted, we're working in K8S here and things are different. But the concept is the same. The user "nobody" can safely be assumed to not be assigned to any user, and is generally designed to be used as an "isolation" user already. We can definitely use other users as needed, but in my mind we're solving a problem we don't really have ... unless we were to decide to keep a catalog of container users and adhere to it very strictly ...

Reply:
@drivera-armedia We should not try to anticipate where the container will run. I think we should not make any assumptions about features of the host OS that will run the container. I can setup a host OS where my user id will be the same number as the `nobody` user of the container. Securing access to volumes from the host OS is not the concern of the container. I totally fail to understand what's the point of keeping a "catalog of container users". User `nobody` or any other user are the same as far as the kernel is concerned, except for user id 0 (and I am paraphrasing what you wrote). 

What's your take on ensuring the containerized app has read-only access to files/directories it only needs to read and write access only to files/directories is has to be able to write to?

Reply:
It's a very beautiful ideal, but it requires two things: a) extra work, and b) perfect certainty that this will not hobble the application in some unforeseen way (i.e. b/c it has a bug that tries to write to an anticipated read-only area).

I see the value, but not the immediate need. Or, more accurately, I see too much effort being required to make sure we can leverage this safely without risking unforeseen consequences - effort that we can better expend elsewhere.